aws/src/main/java/org/apache/iceberg/aws/AssumeRoleSingleRegionFactory.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws;
+
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import software.amazon.awssdk.awscore.client.builder.AwsClientBuilder;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.sts.StsClientBuilder;
+
+/**
+ * Factory that uses the same configured region for all AWS client initialization,
+ * and uses region in the default AWS region provider chain for refreshing assume role credentials.
+ */
+public class AssumeRoleSingleRegionFactory implements AwsClientRegionFactory {
+
+  private Region region;
+
+  @Override
+  public void configure(AwsClientBuilder clientBuilder) {
+    // do not configure STS so it can pick up default credentials
+    if (!(clientBuilder instanceof StsClientBuilder)) {
+      clientBuilder.region(region);
+    }
+  }
+
+  @Override
+  public void initialize(Map<String, String> properties) {
+    String regionStr = properties.get(AwsProperties.CLIENT_ASSUME_ROLE_REGION);
+    Preconditions.checkNotNull(regionStr, "Cannot initialize AssumeRoleSingleRegionFactory with null region");
+    region = Region.of(regionStr);
```

Comment:
RegionStr can be an incorrect value.
We need to check if region is correct.

Reply:
There are special regions that cannot be checked, that is why we only check not null.